components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth/action/PreIssueAccessTokenProcessor.java

```diff
@@ -0,0 +1,719 @@
+                                                        ActionInvocationSuccessResponse
+                                                                actionInvocationSuccessResponse) {
+
+        OAuthTokenReqMessageContext tokenMessageContext =
+                (OAuthTokenReqMessageContext) eventContext.get("tokenMessageContext");
+        PreIssueAccessTokenEvent preIssueAccessTokenEvent = (PreIssueAccessTokenEvent) event;
+        List<PerformableOperation> operationsToPerform = actionInvocationSuccessResponse.getOperations();
+
+        AccessToken requestAccessToken = preIssueAccessTokenEvent.getAccessToken();
+        AccessToken.Builder responseAccessTokenBuilder = preIssueAccessTokenEvent.getAccessToken().copy();
+        List<OperationExecutionResult> operationExecutionResultList = new ArrayList<>();
+
+        if (operationsToPerform != null) {
+            for (PerformableOperation operation : operationsToPerform) {
+                switch (operation.getOp()) {
+                    case ADD:
+                        operationExecutionResultList.add(
+                                handleAddOperation(operation, requestAccessToken, responseAccessTokenBuilder));
+                        break;
+                    case REMOVE:
+                        operationExecutionResultList.add(
+                                handleRemoveOperation(operation, requestAccessToken, responseAccessTokenBuilder));
+                        break;
+                    case REPLACE:
+                        operationExecutionResultList.add(
+                                handleReplaceOperation(operation, requestAccessToken, responseAccessTokenBuilder));
+                        break;
+                    default:
+                        break;
+                }
+            }
+        }
+
+        logOperationExecutionResults(actionType, operationExecutionResultList);
+
+        AccessToken responseAccessToken = responseAccessTokenBuilder.build();
+        updateTokenMessageContext(tokenMessageContext, responseAccessToken);
+
+        return new ActionExecutionStatus(ActionExecutionStatus.Status.SUCCESS, eventContext);
+    }
+
+    private void logOperationExecutionResults(ActionType actionType,
+                                              List<OperationExecutionResult> operationExecutionResultList) {
+
+        //todo: need to add to diagnostic logs
+        if (LOG.isDebugEnabled()) {
+            ObjectMapper objectMapper = new ObjectMapper();
+            try {
+                String executionSummary = objectMapper.writeValueAsString(operationExecutionResultList);
+                LOG.debug(String.format("Processed response for action type: %s. Results of operations performed: %s",
+                        actionType, executionSummary));
+            } catch (JsonProcessingException e) {
+                LOG.debug("Error occurred while logging operation execution results.", e);
```

Comment:
Shall we make this as an error log?

Reply:
This gets executed in debug scope only. So don't see a reason to indicate that as an error. 
Errors should be errors that the system really needs to be alerted on.